ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -472,106 +476,143 @@ mkInternalStateDB :: (HasCallStack, MonadThrow m, MonadCatch m, Ord blockId)
-                    Just _err -> do
-                        -- the handle of the parser is closed at this point. We need to
-                        -- reOpen the file in AppendMode now (parser opens with ReadMode).
-                        -- Note that no file is open at this point, so we can safely open
-                        -- with AppendMode any file, without the fear of opening multiple
-                        -- concurrent writers, which is not allowed.
-
-                        -- TODO(kde) we should add a Warning log here.
-                        withFile hasFS file (AppendMode AllowExisting) $ \hndl ->
-                            hTruncate hndl (fromIntegral offset)
-                        return ()
-                newRevMp <- fromEither err $ reverseMap file revMp fileMp
-                let (fileInfo, maxSlotOfFile) = FileInfo.fromParsedInfo ls
-                let newMp = Map.insert file fileInfo mp
-                let newMaxSlot = maxSlotList $ catMaybes [maxSlot, maxSlotOfFile]
-                let newSuccMp = foldr
-                        (\(_,(_, blockInfo)) pm -> insertMapSet pm (bbid blockInfo, bpreBid blockInfo))
+        go mp revMp succMp maxSlot haveLessThanN (file : restFiles) = do
+            (blocks, mErr) <- parse parser file
+            updateAndGo mp revMp succMp maxSlot haveLessThanN file restFiles blocks mErr
+        go mp revMp succMp _maxSlot haveLessThanN [] = do
+            hndl <- hOpen hasFS fileToWrite (AppendMode AllowExisting)
+            return $ InternalState {
+                  _currentWriteHandle = hndl
+                , _currentWritePath   = fileToWrite
+                , _currentWriteOffset = offset'
+                , _nextNewFileId      = nextNewFileId'
+                , _currentMap         = mp'
+                , _currentRevMap      = revMp
+                , _currentSuccMap     = succMp
+                , _currentMaxSlotNo   = FileInfo.maxSlotInFiles (Index.elems mp')
+            }
+            where
+                (fileToWrite, nextNewFileId', mp', offset') =
+                    case (safeMaximumOn (\(a,_,_) -> a) haveLessThanN, lastFd) of
+                        (Nothing, _) -> newFileInfo mp $ case lastFd of
+                            Just lst -> lst + 1
+                            Nothing -> 0
+                        (_, Nothing) -> error errorMsg
+                        (Just (fd, wrfile, size), Just lst) | fd == lst->
+                            -- if the file with biggest index is non-empty,
+                            -- then this is the file we're looking for.
+                            (wrfile, lst + 1, mp, size)
+                        (_, Just lst) ->
+                            -- If it's not the last file, we just ignore it and
+                            -- open a new one.
+                            newFileInfo mp $ lst + 1
+
+        updateAndGo mp revMp succMp maxSlot haveLessThanN file restFiles blocks mErr = do
```

Comment:
With so many arguments, I think a type signature would help me understand it better.